.github/workflows/ci.yml

```diff
@@ -0,0 +1,180 @@
+name: CI
+
+on: [push, pull_request]
+
+jobs:
+  ci:
+    strategy:
+      # Allows for matrix sub-jobs to fail without canceling the rest
+      fail-fast: false
+
+      # MATRIX:
+      # =======
+      # Required parameters:
+      #  os                  the os to run on
+      #  python-version      the python version to use
+      #  backend             the backend to use
+      #  env                 any additional env variables. Set to '{}' for none
+      matrix:
+        # Tests [amd64]
+        #
+        # FIXME: 'cpp' tests seems to fail due to compilation errors (numpy_pythran_unit)
+        # in 3.6+ and test failures (builtin_float) in 3.5<
+        os: [ubuntu-18.04]
+        backend: [c, cpp]
+        python-version: [2.7, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9]
+        env: [{}]
+
+        include:
+          # Ubuntu sub-jobs:
+          # ================
+          # Linting
+          - os: ubuntu-18.04
+            python-version: 3.7
+            backend: "c,cpp"
+            env: { TEST_CODE_STYLE: 1 }
+          # Limited API
+          - os: ubuntu-18.04
+            python-version: 3.6
+            backend: "c,cpp"
+            env: { LIMITED_API: --limited-api, EXCLUDE: --no-file }
+          - os: ubuntu-18.04
+            python-version: 3.7
+            backend: "c,cpp"
+            env: { LIMITED_API: --limited-api, EXCLUDE: --no-file }
+          - os: ubuntu-18.04
+            python-version: 3.8
+            backend: "c,cpp"
+            env: { LIMITED_API: --limited-api, EXCLUDE: --no-file }
+          # Stackless
+          - os: ubuntu-18.04
+            python-version: 2.7
+            backend: c
+            env: { STACKLESS: true, PY: 3 }
+          - os: ubuntu-18.04
+            python-version: 3.6
+            backend: c
+            env: { STACKLESS: true, PY: 3 }
+          # Pypy
+          - os: ubuntu-18.04
+            python-version: pypy2
+            backend: c
+            env: { PYTHON_VERSION: "pypy" }
+          - os: ubuntu-18.04
+            python-version: pypy3
+            backend: c
+            env: {}
+          # Pypy [allowed-failures] - These specifically test known bugs
+          - os: ubuntu-18.04
+            python-version: pypy2
+            backend: c
+            env:
+              {
+                PYTHON_VERSION: "pypy",
+                EXCLUDE: "--listfile=tests/pypy_bugs.txt --listfile=tests/pypy2_bugs.txt bugs",
+              }
+          - os: ubuntu-18.04
+            python-version: pypy3
+            backend: c
+            env:
+              {
+                EXCLUDE: "--listfile=tests/pypy_bugs.txt --listfile=tests/pypy2_bugs.txt bugs",
+              }
+          # Coverage
+          # FIXME: Disabled due to outputing infinitely
```

Comment:
At least of travis indefinite output is usually a sign that something has segfaulted and the test-running hasn't picked it up (so is forever waiting for it to finish). It's always a bit of a pain to diagnose but just be the last test that was started.

Reply:
The weird thing is that it doesn't just stop outputting, it keeps running tests. I left one of the pipelines run overnight and I woke up to 20k lines of output for some reason.

Reply:
Is the output just a list of timestamps? Because the "watcher" thread will continue to produce those forever. Do you have a link to an output like that?

Reply:
Found one: https://github.com/davfsa/cython/runs/154308827.
#
As I keep running tests, I can also seem to reproduce this on the pipelines that run on diff archs:
`ppc64le`: https://github.com/davfsa/cython/runs/1548634363
`aarch`: https://github.com/davfsa/cython/runs/1548634349

Reply:
It doesn't look like any of them are repeating tests (I'm assuming the first URL you linked to has a 1 on the end...). And they only have run-times of ~1hr. Unless I'm missing something I think they're just running slowly possibly?

Reply:
Well, turns out they dont run infinitely, its just them taking 5h to complete, lol: https://github.com/davfsa/cython/runs/1549995502.

@da-woods Should I keep them disabled or enable them again? (This also goes for the macos one and coverage, as they also take a really long time)

cc @scoder

Reply:
> Should I keep them disabled or enable them again?

Coverage is currently disabled https://github.com/cython/cython/blob/3cc9cdb15469187c63248a480b6291a05671b61e/.travis.yml#L57-L59 so that seems like it should remain disabled.

Don't know about the others. My feeling is that MacOS spends quite a bit of time failing but not a whole lot of time failing in a useful way. But I don't really know

Reply:
Generally this looks like a good direction. You mention that some of the tests take a /long/ time to run, do they also take a long time on jenkins?

Reply:
Don't think any tests get run on Jenkins? But the alternative architecture tests (AMD64/PowerPC) don't look especially slow on Travis.

Reply:
GitHub Actions doesn't allow to run on other architectures (unless you set up your own runners on those architectures). The way to achieve running on different architectures is through an action that uses docker images to simulate the architecture. This seems to have an effect on the compiling of the tests, causing them to take longer (only a speculation). 

The weird this is that this also happens on macOS when using python 3.8, but not on 2.7 (might be a test dependency slowing it down). I will investigate more on this cause might be the script or something not installed correctly causing it (has happened before I opened this pr)

Any info welcome!

Reply:
> Generally this looks like a good direction. You mention that some of the tests take a /long/ time to run, do they also take a long time on jenkins?

Jenkins is self-hosted, so you would need to set up your own instance for that (something which I haven't done). Also, this would have the same problem with different arch's as GitHub actions, since you would need the phisical hardware to run it on.

Reply:
Sorry, I meant to compare runtimes against travis ci. (Yeah, I would rather not get (back) into the business of hosting our own CI if it can be avoided.)